File: macos/Tests/BambuDryCoreTests/MessagesTests.swift
import XCTest
@testable import BambuDryCore

final class MessagesTests: XCTestCase {

    // MARK: - Outbound encode

    func testStartCommandMatchesProtocolShape() throws {
        let req = DryRequest.start(amsId: 1,
                                   filament: "PETG",
                                   tempC: 65,
                                   hours: 8,
                                   sequenceId: "42")
        let data = try JSONEncoder().encode(req)
        let dict = try JSONSerialization.jsonObject(with: data) as! [String: Any]
        let print = dict["print"] as! [String: Any]
        XCTAssertEqual(print["command"] as? String, "ams_filament_drying")
        XCTAssertEqual(print["sequence_id"] as? String, "42")
        XCTAssertEqual(print["ams_id"] as? Int, 1)
        XCTAssertEqual(print["mode"] as? Int, DryCtrlMode.onTime.rawValue)
        XCTAssertEqual(print["filament"] as? String, "PETG")
        XCTAssertEqual(print["temp"] as? Int, 65)
        XCTAssertEqual(print["duration"] as? Int, 8)
        XCTAssertEqual(print["humidity"] as? Int, 0)
        XCTAssertEqual(print["rotate_tray"] as? Bool, false)
        // DryRequest.start defaults coolingTemp to 50°C — Bambu's recommended
        // post-dry cooldown setpoint. Pin to that default so the encoded shape
        // matches what the printer actually receives in production.
        XCTAssertEqual(print["cooling_temp"] as? Int, 50)
        XCTAssertEqual(print["close_power_conflict"] as? Bool, false)
    }

    func testStopCommandMatchesProtocolShape() throws {
        let req = DryRequest.stop(amsId: 0, sequenceId: "7")
        let data = try JSONEncoder().encode(req)
        let dict = try JSONSerialization.jsonObject(with: data) as! [String: Any]
        let print = dict["print"] as! [String: Any]
        XCTAssertEqual(print["command"] as? String, "ams_filament_drying")
        XCTAssertEqual(print["mode"] as? Int, DryCtrlMode.off.rawValue)
        XCTAssertEqual(print["filament"] as? String, "")
        XCTAssertEqual(print["temp"] as? Int, 0)
        XCTAssertEqual(print["duration"] as? Int, 0)
    }

    // MARK: - Bitfield decoding

    func testInfoBitsDecodeRealFieldLayout() {
        // Real printer info "10002003" → bits 0..3 = 3 (N3F model), bits 4..7 = 0 (DryStatus.off).
        XCTAssertEqual(AmsReport.bits(in: "10002003", start: 0, count: 4), 3)
        XCTAssertEqual(AmsReport.bits(in: "10002003", start: 4, count: 4), 0)
        XCTAssertEqual(AmsReport.bits(in: "10002004", start: 0, count: 4), 4)  // N3S
    }

    func testInfoBitsDryingExample() {
        // Synthesised "drying" state: model=N3F (3) in bits 0..3, DryStatus.drying (2) in bits 4..7.
        // 0x23 = 0010 0011
        XCTAssertEqual(AmsReport.bits(in: "23", start: 0, count: 4), 3)
        XCTAssertEqual(AmsReport.bits(in: "23", start: 4, count: 4), 2)
    }

    func testInfoBitsHandlesPrefix() {
        XCTAssertEqual(AmsReport.bits(in: "0x10002003", start: 0, count: 4), 3)
        XCTAssertEqual(AmsReport.bits(in: "0X10002004", start: 0, count: 4), 4)
    }

    func testInfoBitsRejectsNonHex() {
        XCTAssertNil(AmsReport.bits(in: "ZZZ", start: 0, count: 4))
    }

    // MARK: - Snapshot derivation from a real H2D fixture
    // Captured payload: 1× AMS 2 Pro (id=0) + 2× AMS HT (id=128, 129), all idle.

    func testFixtureDecodesAllThreeAmsUnits() throws {
        let url = Bundle.module.url(forResource: "ams_report", withExtension: "json")!
        let data = try Data(contentsOf: url)
        let envelope = try JSONDecoder().decode(ReportEnvelope.self, from: data)
        let amsList = envelope.print?.ams?.ams ?? []
        XCTAssertEqual(amsList.count, 3)
        XCTAssertEqual(amsList.map { $0.id }, ["0", "128", "129"])
    }

    func testFixtureMainAmsIsN3FIdle() throws {
        let ams = try loadAms(at: 0)
        XCTAssertEqual(ams.id, "0")
        XCTAssertEqual(ams.humidity_raw, "24")
        XCTAssertEqual(ams.info, "10002003")

        // Auto-detect model from info bits 0..3
        XCTAssertEqual(ams.amsModel(), .n3f)

        let snap = ams.snapshot()!
        XCTAssertEqual(snap.amsId, 0)
        XCTAssertEqual(snap.model, .n3f)
        XCTAssertEqual(snap.humidityPercent, 24)
        XCTAssertTrue(snap.supportsDrying)
        XCTAssertFalse(snap.isCurrentlyDrying)
        XCTAssertEqual(snap.leftDryTimeMinutes, 0)
        XCTAssertEqual(snap.cannotDryReasons, [])
        XCTAssertEqual(snap.dominantFilamentType, "PLA")
        XCTAssertEqual(snap.currentTempC, 27.2)
    }

    func testFixtureSecondaryAmsAreN3SIdle() throws {
        for index in 1...2 {
            let ams = try loadAms(at: index)
            XCTAssertEqual(ams.amsModel(), .n3s, "ams index=\(index)")
            let snap = ams.snapshot()!
            XCTAssertEqual(snap.model, .n3s)
            XCTAssertEqual(snap.humidityPercent, 28)
            XCTAssertFalse(snap.isCurrentlyDrying)
            XCTAssertEqual(snap.leftDryTimeMinutes, 0)
        }
        XCTAssertEqual(try loadAms(at: 1).id, "128")
        XCTAssertEqual(try loadAms(at: 2).id, "129")
    }

    private func loadAms(at index: Int) throws -> AmsReport {
        let url = Bundle.module.url(forResource: "ams_report", withExtension: "json")!
        let data = try Data(contentsOf: url)
        let envelope = try JSONDecoder().decode(ReportEnvelope.self, from: data)
        return envelope.print!.ams!.ams![index]
    }

    func testSnapshotInvalidHumidity() throws {
        // -1 is the device's "invalid" sentinel.
        let json = """
        {"id":"0","humidity_raw":"-1","temp":"24.5","info":"00","dry_time":0}
        """.data(using: .utf8)!
        let report = try JSONDecoder().decode(AmsReport.self, from: json)
        let snap = report.snapshot(model: .n3f)
        XCTAssertNil(snap.humidityPercent)
        XCTAssertFalse(snap.isCurrentlyDrying)
    }

    func testCannotDryReasonsParsed() throws {
        let json = """
        {"id":"1","humidity_raw":"55","info":"00","dry_time":0,"dry_sf_reason":[2,7]}
        """.data(using: .utf8)!
        let report = try JSONDecoder().decode(AmsReport.self, from: json)
        let snap = report.snapshot(model: .n3s)
        XCTAssertEqual(snap.cannotDryReasons, [.amsBusy, .upgrading])
    }
}

File: windows/tests/BambuDry.Core.Tests/MessagesTests.cs
using System.Text.Json;
using BambuDry.Core;
using BambuDry.Core.Net;

namespace BambuDry.Core.Tests;

public class MessagesTests
{
    private static readonly string FixturePath =
        Path.Combine(AppContext.BaseDirectory, "Fixtures", "ams_report.json");

    // MARK: - Outbound encode

    [Fact]
    public void StartCommandMatchesProtocolShape()
    {
        var req = DryRequest.Start(amsId: 1, filament: "PETG", tempC: 65, hours: 8, sequenceId: "42");
        var json = JsonSerializer.Serialize(req, WireJson.Options);
        var doc = JsonDocument.Parse(json);
        var print = doc.RootElement.GetProperty("print");

        Assert.Equal("ams_filament_drying", print.GetProperty("command").GetString());
        Assert.Equal("42",                  print.GetProperty("sequence_id").GetString());
        Assert.Equal(1,                     print.GetProperty("ams_id").GetInt32());
        Assert.Equal((int)DryCtrlMode.OnTime, print.GetProperty("mode").GetInt32());
        Assert.Equal("PETG",                print.GetProperty("filament").GetString());
        Assert.Equal(65,                    print.GetProperty("temp").GetInt32());
        Assert.Equal(8,                     print.GetProperty("duration").GetInt32());
        Assert.Equal(0,                     print.GetProperty("humidity").GetInt32());
        Assert.False(                       print.GetProperty("rotate_tray").GetBoolean());
        // DryRequest.Start defaults coolingTemp to 50°C — Bambu's recommended
        // post-dry cooldown setpoint. Pin to that default so the encoded shape
        // matches what the printer actually receives in production.
        Assert.Equal(50,                    print.GetProperty("cooling_temp").GetInt32());
        Assert.False(                       print.GetProperty("close_power_conflict").GetBoolean());
    }

    [Fact]
    public void StopCommandMatchesProtocolShape()
    {
        var req = DryRequest.Stop(amsId: 0, sequenceId: "7");
        var json = JsonSerializer.Serialize(req, WireJson.Options);
        var doc = JsonDocument.Parse(json);
        var print = doc.RootElement.GetProperty("print");

        Assert.Equal("ams_filament_drying",   print.GetProperty("command").GetString());
        Assert.Equal((int)DryCtrlMode.Off,    print.GetProperty("mode").GetInt32());
        Assert.Equal("",                      print.GetProperty("filament").GetString());
        Assert.Equal(0,                       print.GetProperty("temp").GetInt32());
        Assert.Equal(0,                       print.GetProperty("duration").GetInt32());
    }

    // MARK: - Bitfield decoding

    [Fact]
    public void InfoBitsDecodeRealFieldLayout()
    {
        // Real printer info "10002003" → bits 0..3 = 3 (N3F), bits 4..7 = 0 (DryStatus.Off).
        Assert.Equal(3, AmsReport.Bits("10002003", start: 0, count: 4));
        Assert.Equal(0, AmsReport.Bits("10002003", start: 4, count: 4));
        Assert.Equal(4, AmsReport.Bits("10002004", start: 0, count: 4));  // N3S
    }

    [Fact]
    public void InfoBitsDryingExample()
    {
        // Synthesised "drying" state: model=N3F (3) bits 0..3, DryStatus.Drying (2) bits 4..7.
        // 0x23 = 0010 0011
        Assert.Equal(3, AmsReport.Bits("23", start: 0, count: 4));
        Assert.Equal(2, AmsReport.Bits("23", start: 4, count: 4));
    }

    [Fact]
    public void InfoBitsHandlesPrefix()
    {
        Assert.Equal(3, AmsReport.Bits("0x10002003", start: 0, count: 4));
        Assert.Equal(4, AmsReport.Bits("0X10002004", start: 0, count: 4));
    }

    [Fact]
    public void InfoBitsRejectsNonHex()
    {
        Assert.Null(AmsReport.Bits("ZZZ", start: 0, count: 4));
    }

    // MARK: - Snapshot derivation from a real H2D fixture
    // Captured payload: 1× AMS 2 Pro (id=0) + 2× AMS HT (id=128, 129), all idle.

    [Fact]
    public void FixtureDecodesAllThreeAmsUnits()
    {
        var envelope = LoadFixture();
        var amsList = envelope.Print?.Ams?.Ams ?? new List<AmsReport>();
        Assert.Equal(3, amsList.Count);
        Assert.Equal(new[] { "0", "128", "129" }, amsList.Select(a => a.Id).ToArray());
    }

    [Fact]
    public void FixtureMainAmsIsN3FIdle()
    {
        var ams = LoadAms(0);
        Assert.Equal("0", ams.Id);
        Assert.Equal("24", ams.HumidityRaw);
        Assert.Equal("10002003", ams.Info);

        Assert.Equal(AmsModel.N3F, ams.GetAmsModel());

        var snap = ams.Snapshot()!;
        Assert.Equal(0, snap.AmsId);
        Assert.Equal(AmsModel.N3F, snap.Model);
        Assert.Equal(24, snap.HumidityPercent);
        Assert.True(snap.SupportsDrying);
        Assert.False(snap.IsCurrentlyDrying);
        Assert.Equal(0, snap.LeftDryTimeMinutes);
        Assert.Empty(snap.CannotDryReasons);
        Assert.Equal("PLA", snap.DominantFilamentType);
        Assert.Equal(27.2, snap.CurrentTempC);
    }

    [Fact]
    public void FixtureSecondaryAmsAreN3SIdle()
    {
        for (var i = 1; i <= 2; i++)
        {
            var ams = LoadAms(i);
            Assert.Equal(AmsModel.N3S, ams.GetAmsModel());
            var snap = ams.Snapshot()!;
            Assert.Equal(AmsModel.N3S, snap.Model);
            Assert.Equal(28, snap.HumidityPercent);
            Assert.False(snap.IsCurrentlyDrying);
            Assert.Equal(0, snap.LeftDryTimeMinutes);
        }
        Assert.Equal("128", LoadAms(1).Id);
        Assert.Equal("129", LoadAms(2).Id);
    }

    [Fact]
    public void SnapshotInvalidHumidity()
    {
        // -1 is the device's "invalid" sentinel.
        const string json = """{"id":"0","humidity_raw":"-1","temp":"24.5","info":"00","dry_time":0}""";
        var report = JsonSerializer.Deserialize<AmsReport>(json, WireJson.Options)!;
        var snap = report.Snapshot(AmsModel.N3F);
        Assert.Null(snap.HumidityPercent);
        Assert.False(snap.IsCurrentlyDrying);
    }

    [Fact]
    public void CannotDryReasonsParsed()
    {
        const string json = """{"id":"1","humidity_raw":"55","info":"00","dry_time":0,"dry_sf_reason":[2,7]}""";
        var report = JsonSerializer.Deserialize<AmsReport>(json, WireJson.Options)!;
        var snap = report.Snapshot(AmsModel.N3S);
        Assert.Equal(new[] { CannotDryReason.AmsBusy, CannotDryReason.Upgrading }, snap.CannotDryReasons);
    }

    private static ReportEnvelope LoadFixture()
    {
        var bytes = File.ReadAllBytes(FixturePath);
        return JsonSerializer.Deserialize<ReportEnvelope>(bytes, WireJson.Options)!;
    }

    private static AmsReport LoadAms(int index) => LoadFixture().Print!.Ams!.Ams![index];
}
